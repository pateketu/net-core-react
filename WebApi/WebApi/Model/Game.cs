using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Newtonsoft.Json.Converters;

namespace WebApi.Model
{
    public class Game
    {
        public Guid Id { get; set; }
        [JsonConverter(typeof(StringEnumConverter))]
        public Level Level { get; set; }
        public int A { get; set; }
        public int B { get; set; }
        public int TimeFrameSeconds { get; set; }
        [JsonIgnore]
        public DateTime StartedAt { get; set; }
        [JsonIgnore]
        public int NumOfConsecutiveCorrectAnswers { get; set; }

        // Method to create an exercise of type a + b = ?
        public static Game CreateGame(int a, int b, int defaultTimeFrameSeconds)
        {
            return new Game() { Id = Guid.NewGuid(), A = a, B = b, TimeFrameSeconds = defaultTimeFrameSeconds, StartedAt = DateTime.Now };
        }
    }
}
