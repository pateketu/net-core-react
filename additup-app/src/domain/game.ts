export default class Game {
   public id: string = '';
   public level: string = '';
   public a: number = 0;
   public b: number = 0;
   public timeFrameSeconds: number = 0;
}
