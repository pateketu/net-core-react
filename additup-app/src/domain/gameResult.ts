import Game from './game';

export class GameResult {
    public inCorrectAnswer: boolean = false;
    public game: Game | undefined;
    public allLevelsFinished: boolean = false;
}
