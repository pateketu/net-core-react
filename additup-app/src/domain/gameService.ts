
import Game from './game';
import { Promise, async } from 'q';
import { GameResult } from './gameResult';

export async function start() {
    return Promise<Game>((resolve) => {
        resolve(new Game());
    });
}

export async function answer(ans: number) {
    return Promise<GameResult>((resolve) => {
        resolve(new GameResult());
    });
}
