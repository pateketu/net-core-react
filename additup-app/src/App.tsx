import React, { Component } from 'react';
import './App.css';
import Answer from './components/Answer';
import Timer from './components/Timer';
import Game from './domain/game';
import * as gameService from './domain/gameService';

interface IAppState {
  expired: boolean;
  game?: Game;
}

class App extends Component<any, IAppState> {

  constructor(props: any) {
    super(props);
    this.state = { expired: false};
  }

  public async componentDidMount() {
    const game = await gameService.start();
    this.setState({game});
  }

  public render() {
    if (!this.state.game) {
      return <React.Fragment>
              {this.header()}
              <div>Loading...</div>
          </React.Fragment>;
    }
    return (
      <div className="App">
        {this.header()}
        {!this.state.expired  ?
              this.equation(this.state.game)
            : this.reStartButton()
        }
      </div>
    );
  }

  private header() {
    return  <h1>Add It Up!</h1>;
  }
  private equation(game: Game) {
    return <Timer seconds={game.timeFrameSeconds}
                  onExpired={this.onExpired}>
          <b>Your current level is {game.level}</b>
          <Answer a={game.a} b={game.b} onAnswer={(answer: number) => {
            // Post the answer to
          }}> </Answer>

      </Timer>;
  }

  private reStartButton() {
    return <div><button onClick={() => this.setState({expired: false})}>Re-Start</button></div>;
  }

  private onExpired = () => {
    this.setState({expired: true});
  }
}

export default App;
