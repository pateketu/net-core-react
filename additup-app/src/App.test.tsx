import React from 'react';
import App from './App';
import {mount, shallow} from 'enzyme';
import * as gameService from './domain/gameService';
import { Promise, Deferred } from 'q';
import Game from './domain/game';
import toJson from 'enzyme-to-json';

describe('Game App', () => {
  it('Renders game', (done) => {
      const mockStart = jest.spyOn(gameService, 'start');
      const fakeGame = {...new Game(), a: 6, b: 6, timeFrameSeconds: 20, level: 'Beginner'};

      const promise = Promise((r) => r(fakeGame));
      mockStart.mockReturnValue(promise);

      const wrapper = mount(<App/>);

      promise.then(() => {
         setImmediate(() => {
          wrapper.update();
          expect(toJson(wrapper, {noKey: true})).toMatchSnapshot();
          done();
        });
      }, () => {/*Reject*/});
  });
});
