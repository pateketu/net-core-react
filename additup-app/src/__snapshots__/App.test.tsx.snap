// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game App Renders game 1`] = `
<App>
  <div
    className="App"
  >
    <h1>
      Add It Up!
    </h1>
    <Timer
      onExpired={[Function]}
      seconds={20}
    >
      <div>
        20
         Seconds Remaining
      </div>
      <b>
        Your current level is 
        Beginner
      </b>
      <Answer
        a={6}
        b={6}
        onAnswer={[Function]}
      >
        <div>
          <AnswerInput
            a={6}
            b={6}
            onAnswer={[Function]}
          >
            <div>
              <div
                className="equation"
              >
                <h4>
                  6
                </h4>
                <h4>
                  +
                </h4>
                <h4>
                  6
                </h4>
                <div>
                  <input
                    onChange={[Function]}
                    type="text"
                  />
                </div>
              </div>
            </div>
          </AnswerInput>
          <button
            disabled={true}
            onClick={[Function]}
          >
            Asnwer
          </button>
        </div>
      </Answer>
    </Timer>
  </div>
</App>
`;
